### PR TITLE
DrivAerML: Fourier+no-EMA+T_max=30 long training (most critical)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current DrivAerML best (51.35% val) was achieved in only 2 epochs with a 30-min budget. The training loss was still steeply descending at cutoff. With the 180-min budget and `--epochs 999`, we expect 12–15 epochs — potentially reaching the 30–40% range. This is the highest-priority DrivAerML experiment right now.

On AirfRANS, extending from 2→8 epochs with the same config yielded a 17.4% relative improvement. A similar multiplier on DrivAerML (2→12+ epochs) could push us from ~51% into the 30–40% territory.

## Instructions

Run the following command — do NOT modify train.py:

```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine_t_max 30 \
  --no-use-ema \
  --enable-fourier \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --wandb_group historia-drivaerml-fourier-longrun
```

**Budget:** 180 min — let it run to the full timeout.

**Report (in a PR comment when done):**
- val_primary/surface_rel_l2_pct and test_primary/surface_rel_l2_pct at each epoch
- Total epochs completed
- Per-epoch trajectory (note at what epoch training starts to plateau, if at all)
- W&B run ID

## Baseline

- **Primary metric:** val_primary/surface_rel_l2_pct (lower is better)
- **Current best val:** 51.35%
- **Current best test:** 52.06%
- **Best PR:** #2475 (chihiro — Fourier + no-EMA, T_max=30, AdamW lr=5e-4, only 2 epochs)
- **External target:** <3.71% (AB-UPT, ~500 epochs)
- **W&B run:** 5ncrjm32

**CRITICAL:** Always pass `--epochs 999` — the default is 2 (hardcoded bug). Without this flag, training stops after 2 epochs regardless of budget.

**DrivAerML standard flags** (required to avoid OOM): `--batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2499 because Superseded: historia reassigned to #2562 (Kutta TE constraint).

**Comment by tcapelle** [comment]
STUDENT historia:

## Results

### Key Metrics

| Metric | Baseline (PR #2475) | This Run (Best) | This Run (Final, Ep 32) | Improvement |
|--------|---------------------|-----------------|-------------------------|-------------|
| val_primary/surface_rel_l2_pct | 51.35% | **18.81%** (ep 29) | 21.92% | **63.4% relative** |
| test_primary/surface_rel_l2_pct | 52.06% | — | **23.11%** | **55.6% relative** |

### Per-Epoch Trajectory

| Epoch | Val % | Loss | LR | Notes |
|-------|-------|------|----|-------|
| 1 | 61.71 | 0.828 | 2.16e-05 | Warmup |
| 2 | 54.10 | 0.568 | 4.17e-04 | Near baseline |
| 3 | 43.62 | 0.425 | 1.73e-04 | First big drop |
| 4 | 40.79 | 0.342 | 2.24e-04 | |
| 5 | 38.13 | 0.296 | 3.75e-04 | |
| 6 | 35.55 | 0.263 | 4.77e-05 | Cosine trough — new best |
| 7 | 37.36 | 0.227 | 4.95e-04 | LR peak regression |
| 8 | 30.14 | 0.206 | 5.46e-06 | Trough — broke 30% |
| 9 | 31.59 | 0.188 | 4.52e-04 | LR peak |
| 10 | 28.04 | 0.172 | 1.25e-04 | New best |
| 11 | 30.01 | 0.157 | 2.76e-04 | |
| 12 | 27.78 | 0.149 | 3.27e-04 | New best mid-cycle |
| 13 | 25.85 | 0.141 | 8.27e-05 | New best |
| 14 | 27.74 | 0.138 | 4.78e-04 | LR peak |
| 15 | 25.16 | 0.130 | 0.00 | Full cosine cycle (T_max=30) — new best |
| 16 | 28.49 | 0.126 | 4.78e-04 | New cycle start |
| 17 | 23.41 | 0.121 | 8.27e-05 | New best |
| 18 | 25.77 | 0.115 | 3.27e-04 | |
| 19 | 23.65 | 0.116 | 2.76e-04 | |
| 20 | 22.81 | 0.114 | 1.25e-04 | New best |
| 21 | 25.79 | 0.105 | 4.52e-04 | |
| 22 | 22.09 | 0.103 | 5.46e-06 | Trough — new best |
| 23 | 25.58 | 0.100 | 4.95e-04 | |
| 24 | 20.88 | 0.098 | 4.77e-05 | Trough — new best |
| 25 | 22.62 | 0.094 | 3.75e-04 | |
| 26 | 22.58 | 0.090 | 2.24e-04 | |
| 27 | 23.92 | 0.087 | 1.73e-04 | |
| 28 | 22.27 | 0.086 | 4.17e-04 | |
| 29 | **18.81** | 0.090 | 2.16e-05 | **Trough — BEST VAL** |
| 30 | 20.30 | 0.078 | 5.00e-04 | New cycle peak |
| 31 | 19.36 | 0.080 | 2.16e-05 | Trough — near best |
| 32 | 21.92 | 0.075 | 4.17e-04 | Final epoch (timeout) |

### Training Details

- **Total epochs completed:** 32
- **Budget used:** 180 min (full timeout)
- **W&B run ID:** `6htqewa3` ([link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/6htqewa3))
- **Peak memory:** Not logged in output (GPU 0 used exclusively via CUDA_VISIBLE_DEVICES=0)
- **Command used:**
  ```bash
  cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
    --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
    --no-use-ema --enable-fourier --epochs 999 --batch-size 1 \
    --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
    --max-train-batches 394 --max-eval-batches 200 \
    --wandb-group historia-drivaerml-fourier-longrun \
    --agent historia --wandb-name "historia/drivaerml-fourier-noema-longrun"
  ```

**Note:** The PR command had underscores in flag names (`--cosine_t_max`, `--wandb_group`, `--wandb_name`) but argparse required hyphens. Fixed to `--cosine-t-max`, `--wandb-group`, `--wandb-name`.

### What Happened

**The hypothesis is strongly confirmed.** Extended training from 2 → 32 epochs produced a massive improvement:
- Val: 51.35% → 18.81% (63.4% relative improvement)
- Test: 52.06% → 23.11% (55.6% relative improvement)

The training curve shows a clear **cosine-with-warm-restarts** pattern — each LR trough produces the best validation metrics, with temporary regressions at LR peaks. Best results consistently occur at cosine cycle troughs.

**No plateau observed** — the model was still improving at the 180-min cutoff. Loss was still descending (0.828 → 0.075) and the per-cycle val improvements were still meaningful (epoch 22: 22.09% → epoch 29: 18.81%).

The val/test gap (18.81% vs 23.11%) suggests mild overfitting, possibly because test evaluation happened at epoch 32 (not epoch 29 where best val occurred). Checkpointing at the best val epoch could reduce this gap.

### Suggested Follow-ups

1. **Even longer training** — the model shows no plateau at 32 epochs. A 360-min budget could reach 60+ epochs and potentially break below 15%.
2. **Checkpoint at best val** — save and evaluate at LR troughs to avoid test regression from LR peak evaluation.
3. **Longer cosine period** — `T_max=60` would give slower, smoother annealing over more epochs, potentially avoiding the mid-cycle val spikes.
4. **Lower peak LR in later cycles** — the warm restart LR jumps cause ~3-4% val regressions. A decaying warm restart schedule could smooth this out.
5. **EMA revisit** — with 32+ epochs of training, EMA might now help smooth the trajectory and improve final metrics.
